### PR TITLE
Add median to assessment question statistics

### PR DESCRIPTION
**Pull request by pmahfuz**
# Median added to statistics table

**Authors:**

- pmahfuz
- nataliaqmc
- arthurfc12

## 1 Executive Summary

Implement a method to check and display the median of a question in the question statistics tab.

## 2 Motivation

The proposal is being made for two reasons:

adding a piece of data deemed to be important when evaluating student performance, being able to check if the mean is an accurate representation of class performance or if there are noticeable discrepancies within the class;
being a kickstarter for future implementations in the statistics tabs by the authors (pmahfuz, nataliaqmc, arthurfc12), who will attempt to make changes in visualization and methods analyzed in the statistics tabs.

## 3 Proposed Implementation

Adding the method for understanding median, instructors will be able to have further knowledge on overall performance in regards to a question, assessment or even course instance. By placing the value of the median next to the value for mean, an easy analysis of discrepancy between the two will be possible, enabling a quick and easy way for instructors to gain a deeper insight between the twos correlation in regards to different questions. The image below illustrates how the change will appear visually:

Statisticstable

## 4 Metrics & Dashboards

Being only an incremental change on a table which already exists, no meaningful metrics will be changes in the PrairieLearn platform, the only difference is de display of a new metric in the statistics table.

## 5 Drawbacks

If the team decides that the table will become too cluttered with the addition of yet another method to the already existing ones, implementation may not be optimal.

## 6 Alternatives

An alternative to the proposal is the usage of a different method when attempting to visualize the median, such as a Mean x Median graph

## 7 Potential Impact and Dependencies

The parties impacted by the above implementation will be instructors because they will be able to easily check a new method for analyzing data. To a lesser degree, the software development team might be impacted by the addition of the code, but, by ensuring that we used proper guidelines in code implementation, we deem the impact to be minimal.

## 8 Unresolved questions

Further iterations in data display are not being addressed in this pull request.

## 9 Conclusion

In conclusion, median visualization will help instructors further understand students' performance by question, enabling the instructors to iterate the questions being given in the course throughout time.

### CODE REVIEW

File: apps/prairielearn/src/lib/assessment.sql [hunk 1]
```diff
@@ -923,6 +923,8 @@ WITH
           avg(question_stats_by_user_or_group.score_perc)
         )
       ) AS mean_question_score,
+      -- Adding median_question_score to the statistics
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/assessment.sql:926]: None of these comments are necessary in any of these files.

File: apps/prairielearn/src/pages/shared/assessmentStatDescriptions.js [hunk 2]
```diff
@@ -4,6 +4,13 @@ var stat_descriptions = {
     non_html_title: 'Mean',
     description: 'Mean score of a question is the average score for all students on the question.',
   },
+  // adding median score 
+  MEDIAN_SCORE: {
+    title: 'Median',
+    non_html_title: 'Median',
+    description:
+      'Median score of a question is the score which separates the lower half and the upper half of students scores.',
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/shared/assessmentStatDescriptions.js:12]: We can bikeshed the best description here; here's a proposal:

```suggestion
      'Median score of a question is the 50th percentile score for all students on the question.',
```

**Review by nwalters512 — COMMENTED** [review]
You'll need to add a database migration to create the new column. See documentation here: https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/migrations/README.md

**Comment by arthurfc12** [comment]
I have reviewed and hereby sign the CLA

**Comment by nataliaqmc** [comment]
I have reviewed and hereby sign the CLA

File: database/tables/course_requests.pg
[the remark this answers is not in the document]
Comment (pmahfuz) [reply database/tables/course_requests.pg:1]: These were the files that had changes made to them after running  `make update-database-description`. I had understood they were supposed to be added to the commit in order to fix the previous errors. I am confused as to what should have been change by this command.

File: database/tables/assessment_questions.pg [hunk 3]
```diff
@@ -31,6 +31,7 @@ columns
     max_submission_score_hist: double precision[]
     max_submission_score_variance: double precision
     mean_question_score: double precision
+    median_question_score: numeric
```
[the remark this answers is not in the document]
Comment (pmahfuz) [reply database/tables/assessment_questions.pg:34]: That might be related to the SQL generation of the statistic, the problem with the values not being shown on the table was related to the type of variable, which was fixed by using another function when "printing" them. Will try and make it a double precision like the rest of the statistics columns.

File: database/tables/course_requests.pg
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply database/tables/course_requests.pg:1]: I believe the easier way to handle this is to merge your PR with master, then run the command again.

[the remark this answers is not in the document]
Comment (pmahfuz) [reply database/tables/course_requests.pg:1]: Would it be a problem if I copied the versions of these files that are on github and submit them in the PR? They would go back to being the way they were

File: database/tables/assessment_questions.pg [hunk 4]
```diff
@@ -31,6 +31,7 @@ columns
     max_submission_score_hist: double precision[]
     max_submission_score_variance: double precision
     mean_question_score: double precision
+    median_question_score: double precision
```
Comment (nwalters512) [inline database/tables/assessment_questions.pg:34]: One final note: you'll need to add this column to `AssessmentQuestionSchema` in `apps/prairielearn/src/lib/db-types.ts`.

**Review by nwalters512 — APPROVED** [review]
